src/Tactic/Premises.agda

```diff
@@ -0,0 +1,294 @@
+
+  -- (3) unbundle hypotheses (always a single product due to `∙`)
+  --     * TODO: generalise to curried rules (which use `→`)
+  unbundleHypotheses : List Type → List Type
+  unbundleHypotheses = concatMap go where go = λ where
+    (def (quote _×_) (hArg _ ∷ hArg _ ∷ vArg x ∷ vArg ys ∷ [])) →
+      x ∷ go ys
+    t → [ t ]
+
+  -- (4) select only the hypotheses that are decidable
+  isDecidable? : AbsTelescope → Type → TC Bool
+  isDecidable? tel ty = runSpeculative $ (_, false) <$>
+    catchTC (checkType (quote it ∙) (tyView (tel , quote Dec ∙⟦ ty ⟧)) >> return true)
+            (return false)
+
+  -- (5) bundle the (decidable) hypotheses back together in a single product
+  bundleHypotheses : List Type → Type
+  bundleHypotheses = λ where
+    [] → quote ⊤ ∙
+    (h ∷ []) → h
+    (h ∷ hs@(_ ∷ _)) → quote _×_ ∙⟦ h ∣ bundleHypotheses hs ⟧
+
+  -- (6) minimize the implicit arguments to only the ones that are actually used in (5)
+  removeUnusedImplicits : AbsTelescope → Type → TC (AbsTelescope × Type)
+  removeUnusedImplicits is ty = go (pred $ length is) is ty
+    where
+    go : ℕ → AbsTelescope → Type → TC (AbsTelescope × Type)
+    go _ []       t = return ([] , t)
+    go n (i ∷ is) t =
+      if L.any (_≡ᵇ n) (freeVars ty) then -- relevant hypothesis
+        map₁ (i ∷_) <$> go (pred n) is t
+      else -- irrelevant hypothesis
+        go (pred n) is (mapFreeVars pred n t)
+
+-- ** Tests.
+private
+
+  -- * The tactics works for one or zero hypotheses.
+  data ℝ : ℕ → ℕ → Set where
+    base :
+      ────────────────────────────────
+      ℝ 0 0
+
+    step : ∀ {n m} →
+      ∙ ℝ n m
+      ────────────────────────────────
+      ℝ (suc n) (suc m)
+
+  module _ ⦃ _ : Dec² ℝ ⦄ where
+    unquoteDecl ℝ-base-premises = genPremises ℝ-base-premises (quote ℝ.base)
+    unquoteDecl ℝ-step-premises = genPremises ℝ-step-premises (quote ℝ.step)
+
+    _ : ℝ-base-premises .proj₁
+    _ = tt
+
+    _ : ℝ-base-premises .proj₂ ≡ yes tt
+    _ = refl
+
+    module _ (n m : ℕ) (𝕣 : ℝ n m) where
+      _ : ℝ-step-premises .proj₁
+      _ = 𝕣
+
+      _ : Dec (ℝ n m)
+      _ = ℝ-step-premises .proj₂
+
+  open import Class.Monoid.Core
+  open import Class.Monoid.Instances
+
+  -- * The tactic works for multiple *decidable* hypotheses.
+  data ℚ {A : Set} ⦃ _ : Monoid A ⦄ : A → A → Set where
+    base :
+      ────────────────────────────────
+      ℚ ε ε
+
+    step : ∀ {n m n′ m′} →
+      ∙ ℚ n m
+      ∙ ℚ n′ m′
+      ────────────────────────────────
+      ℚ (n ◇ n′) (m ◇ m′)
+
+  module _ ⦃ _ : ∀ {A} ⦃ _ : Monoid A ⦄ → Dec² (ℚ {A}) ⦄ where
+    unquoteDecl ℚ-base-premises = genPremises ℚ-base-premises (quote ℚ.base)
+    unquoteDecl ℚ-step-premises = genPremises ℚ-step-premises (quote ℚ.step)
+
+    _ : ℚ-base-premises .proj₁
+    _ = tt
+
+    module _ {A} ⦃ _ : Monoid A ⦄ (n m n′ m′ : A) (𝕢ˡ : ℚ n m) (𝕢ʳ : ℚ n′ m′) where
+      _ : ℚ-step-premises .proj₁
+      _ = 𝕢ˡ , 𝕢ʳ
+
+  -- * The tactic omits *undecidable* hypotheses.
```

Comment:
If we require that all hypotheses are decidable, we then cannot use it in our inference rules (UTXO,UTXOW,etc..), the reason being the recursive call to other rules (e.g. UTXOW has a UTXO step as the last hypothesis).

Reply:
Ah, but the recursive calls are decidable as well! See `Computational⇒Dec` in `Interface.ComputationalRelation`. I think if we turned that into an instance it should work.

Reply:
Are you sure that we would be able to write the computational proofs then as one-liners? Sure it will work for UTXO, maybe for UTXOW, but will it work for everything else?

I am worried that we are replaying the unfortunate scenario of the `deriveComp` tactic, i.e. trying to be too general but doomed to fail in complex cases (like the proof that RATIFY is computational).

Reply:
I'm playing around with it a bit and this does indeed work, but the pattern needs to be changed a bit. I'll investigate more tomorrow.

What we could do is leave the behaviour like it is for now, and if we discover that this works indeed nicely for all computational proofs we can then modify the behaviour to give the stricter guarantees. This works just fine, since if you do make a `Dec` instance for a STS it then doesn't skip it.